lusSTR/marker.py

```diff
@@ -0,0 +1,727 @@
+        back to the UAS region. If the sequence has already been run through UAS, no trimming is
+        required.
+        '''
+        if self.uas:
+            return self.sequence
+        front, back = self._uas_bases_to_trim()
+        if back == 0:
+            back = None
+        else:
+            back *= -1
+        return self.sequence[front:back]
+
+    @property
+    def uas_sequence(self):
+        '''Sequence from the UAS region, matching the orientation of the UAS output
+
+        The UAS software outputs the reverse complement of the forward sequence for some loci.
+        '''
+        if self.data['ReverseCompNeeded'] == "Yes":
+            return reverse_complement(self.forward_sequence)
+        return self.forward_sequence
+
+    @property
+    def flankseq_5p(self):
+        if self.uas:
+            return None
+        front, back = self._uas_bases_to_trim()
+        return self.sequence[:front]
+
+    @property
+    def flank_5p(self):
+        if self.uas:
+            return None
+        flank_rev = collapse_repeats_by_length(self.flankseq_5p[::-1], self.repeat_size)
+        flank = flank_rev[::-1]
+        return flank
+
+    @property
+    def flankseq_3p(self):
+        if self.uas:
+            return None
+        front, back = self._uas_bases_to_trim()
+        if back == 0:
+            return ''
+        return self.sequence[-back:]
+
+    @property
+    def flank_3p(self):
+        if self.uas:
+            return None
+        return collapse_repeats_by_length(self.flankseq_3p, self.repeat_size)
+
+    @property
+    def canonical(self):
+        '''Canonical STR allele designation'''
+        n = self.repeat_size
+        nsubout = self.data['BasesToSubtract']
+        if nsubout == 0:
+            nsubout = None
+        else:
+            nsubout *= -1
+        new_seq = self.uas_sequence[:nsubout]
+        if len(new_seq) % n == 0:
+            canon_allele = int(len(new_seq) / n)
+        else:
+            allele_int = int(len(new_seq) / n)
+            allele_dec = int(len(new_seq) % n)
+            canon_allele = f'{allele_int}.{allele_dec}'
+        return canon_allele
+
+    @property
+    def cannot_split(self):
+        return self.locus in [
+            'D19S433', 'D6S1043', 'TH01', 'D21S11', 'D1S1656', 'D7S820', 'D5S818', 'D12S391',
+            'D9S1122', 'PentaE'
+        ]
+
+    @property
+    def must_split(self):
+        return self.locus in ['D13S317', 'D18S51']
+
+    @property
+    def split_compatible(self):
+        return self.must_split or len(self.uas_sequence) % self.repeat_size == 0
+
+    @property
+    def do_split(self):
+        return not self.cannot_split or self.split_compatible
+
+    @property
+    def annotation(self):
+        bylength = (
+            (self.data['ReverseCompNeeded'] == 'Yes' and not self.cannot_split)
+            or self.locus == 'D16S539'
+        )
```

Comment:
I'll try adding the D3 locus here.

Reply:
That didn't work. It fixed this case, but messed up other cases in the `test_flank_anno` test.